raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java

```diff
@@ -0,0 +1,132 @@
+
+import com.typesafe.config.Config;
+import java.time.Duration;
+import java.time.Instant;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.processor.Cancellable;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.PunctuationType;
+import org.apache.kafka.streams.processor.To;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+import org.hypertrace.core.datamodel.RawSpan;
+import org.hypertrace.core.datamodel.RawSpans;
+import org.hypertrace.core.datamodel.StructuredTrace;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Receives spans keyed by trace_id and stores them. A {@link TraceEmitPunctuator} is scheduled to
+ * run after the {@link RawSpansGroupingTransformer#groupingWindowTimeoutMs} interval to emit the trace. If
+ * any spans for the trace arrive within the {@link RawSpansGroupingTransformer#groupingWindowTimeoutMs}
+ * interval then the {@link RawSpansGroupingTransformer#groupingWindowTimeoutMs} will get reset and the
+ * trace will get an additional {@link RawSpansGroupingTransformer#groupingWindowTimeoutMs} time to accept
+ * spans.
+ */
+public class RawSpansGroupingTransformer implements
+    Transformer<String, RawSpan, KeyValue<String, StructuredTrace>> {
+
+  private static final Logger logger = LoggerFactory.getLogger(RawSpansGroupingTransformer.class);
+  private ProcessorContext context;
+  private KeyValueStore<String, ValueAndTimestamp<RawSpans>> inflightTraceStore;
+  private KeyValueStore<String, Long> traceEmitTriggerStore;
+  private long groupingWindowTimeoutMs;
+  private To outputTopic;
+
+  @Override
+  public void init(ProcessorContext context) {
+    this.context = context;
+    this.inflightTraceStore = (KeyValueStore<String, ValueAndTimestamp<RawSpans>>) context
+        .getStateStore(INFLIGHT_TRACE_STORE);
+    this.traceEmitTriggerStore = (KeyValueStore<String, Long>) context
+        .getStateStore(TRACE_EMIT_TRIGGER_STORE);
+    this.groupingWindowTimeoutMs = ((Config) (context.appConfigs().get(RAW_SPANS_GROUPER_JOB_CONFIG)))
+        .getLong(SPAN_GROUPBY_SESSION_WINDOW_INTERVAL_CONFIG_KEY) * 1000;
+    this.outputTopic = To.child(OUTPUT_TOPIC_PRODUCER);
+    restorePunctuators();
+  }
+
+  @Override
+  public KeyValue<String, StructuredTrace> transform(String key, RawSpan value) {
+    ValueAndTimestamp<RawSpans> rawSpans = inflightTraceStore.get(key);
+    RawSpans agg = rawSpans != null ? rawSpans.value() : RawSpans.newBuilder().build();
+    // add the new span
+    agg.getRawSpans().add(value);
+
+    long currentTimeMs = System.currentTimeMillis();
+
+    if (logger.isDebugEnabled()) {
+      logger.debug("Updating ts=[{}] for trace_id=[{}]", Instant.ofEpochMilli(currentTimeMs), key);
```

Comment:
why not instantiate the ValueAndTimestamp above here and log it instead of making an Instant. does ValueAndTimestamp toString not work?